### PR TITLE
Added an expression to get/set the sneaking state

### PR DESCRIPTION
**Pull request by DeltaRays**
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** none

### CODE REVIEW

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+import ch.njol.skript.classes.Changer;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Sneaking Stat")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:35]: This should be "state"

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+@Description("Whether the player(s) are sneaking")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:36]: This sentence should end with a period.

Comment (ShaneBeee) [reply src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:36]: Also if im not mistaken, you can not force a player to sneak.
Has this been tested? If you can not actually force the player to sneak this should be documented in the description.

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java (hunk 1 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:35]: 👍

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java (hunk 2 above)
Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:36]: It doesn't force them to sneak, it just sets whether or not they appear to be sneaking. I'll add that to the description

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+	
+	@Override
+	protected String getPropertyName() {
+		return "sneaking state";
+	}
+	
+	@Override
+	public Class<Boolean> getReturnType() {
+		return Boolean.class;
+	}
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:74]: As per the coding conventions:
```
SimplePropertyExpression: -> (init) -> convert -> (acceptChange) -> (change) -> getReturnType -> getPropertyName
```
getProperName() should be last

Comment (DeltaRays) [reply src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:74]: Oops

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+import ch.njol.skript.classes.Changer;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Sneaking State")
+@Description("Whether the player(s) are sneaking. Note that it doesn't force them to sneak, it just makes them appear to be / not be sneaking.")
```
Comment (Pikachu920) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:36]: maybe the documentation can say something like "it just changes how other players see the players

**Review by Pikachu920 — DISMISSED** [review]
there's already a `is sneaking` condition, so an effect would make more since here like `make %players appear to be sneaking`

**Comment by FranKusmiruk** [comment]
> there's already a `is sneaking` condition, so an effect would make more since here like `make %players appear to be sneaking`

Considering there's a PR for the toggle changer, I believe we can just forget about these kind of effects for now. `toggle sneak mode of player` looks good enough to me

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+import ch.njol.skript.classes.Changer;
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Sneaking State")
+@Description("Whether the player(s) are sneaking. Note that it doesn't force them to sneak, it just changes how other players see the players")
```
Comment (ShaneBeee) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:36]: Missing period at end of sentence.

**Review by ShaneBeee — DISMISSED** [review]
Other than this one teeny tiny change, looks good to me.

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+import ch.njol.skript.doc.Description;
+import ch.njol.skript.doc.Examples;
+import ch.njol.skript.doc.Name;
+import ch.njol.skript.doc.Since;
+import ch.njol.skript.expressions.base.SimplePropertyExpression;
+import ch.njol.util.coll.CollectionUtils;
+
+@Name("Sneaking State")
+@Description("Whether the player(s) are sneaking. Note that it doesn't force them to sneak, it just changes how other players see the players.")
+@Examples({"set sneak mode of player to true", "send \"%sneaking state of all players%\""})
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:37]: Can you put each example on its own line?

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+	
+	@Override
+	public Boolean convert(final Player player) {
+		return player.isSneaking();
+	}
+	
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(Changer.ChangeMode mode) {
+		if (mode == Changer.ChangeMode.SET || mode == Changer.ChangeMode.RESET) {
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:53]: Nitpick, but you don't *need* the curly brackets

File: src/main/java/ch/njol/skript/expressions/ExprSneakingState.java [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+	
+	@Override
+	public Boolean convert(final Player player) {
+		return player.isSneaking();
+	}
+	
+	@Override
+	@Nullable
+	public Class<?>[] acceptChange(Changer.ChangeMode mode) {
+		if (mode == Changer.ChangeMode.SET || mode == Changer.ChangeMode.RESET)
```
Comment (APickledWalrus) [inline src/main/java/ch/njol/skript/expressions/ExprSneakingState.java:54]: ChangeMode should be imported directly

**Review by APickledWalrus — CHANGES_REQUESTED** [review]
One thing I missed

**Review by APickledWalrus — DISMISSED** [review]
LGTM

**Review by ShaneBeee — APPROVED** [review]

**Review by ShaneBeee — APPROVED** [review]

**Review by APickledWalrus — APPROVED** [review]

**Comment by ShaneBeee** [comment]
I wanted to come back to this one, and question it.
Since you cant actually force the player to sneak, I'm wondering what the point of this really is?

There is already an `is sneaking` condition... personally I see no point to this expression.

**Comment by DeltaRays** [comment]
To make it look like they're sneaking for other players, which could be used in some gamemodes

**Pull request closed without merging**